### PR TITLE
Rebasing tpu branch on a more recent fairseq upstream commit

### PR DESCRIPTION
**Pull request by taylanbil**
This rebase will enable working on RoBERTa.

Transformer on WMT18 performance is about the same. Convergence is verified via bleu score being 25.83 in epoch 5 (will not merge before verifying e2e convergence) and validation loss going down similarly to current tpu branch

Getting this out there for review as it's a big pr. My commit is the last 2, and second to last one is a giant commit. it's many commits squashed to 1.

### CODE REVIEW

**Comment by taylanbil** [comment]
https://github.com/pytorch-tpu/fairseq/pull/19 is an equivalent pr that doesn't have merge conflicts.
